### PR TITLE
Remove Mods enum.

### PR DESCRIPTION
**Pull request by peppy**

### CODE REVIEW

File: osu.Game/Modes/Mod.cs [hunk 1]
```diff
@@ -35,234 +35,129 @@ public abstract class Mod
...
         public override double ScoreMultiplier => 0;
-        public override bool Ranked => false;
-        public override Mods[] DisablesMods => new Mods[] { Mods.Relax, Mods.Autopilot, Mods.SpunOut, Mods.SuddenDeath, Mods.Perfect };
+        public override Type[] IncompatibleMods => new[] { typeof(ModRelax), typeof(ModSuddenDeath), typeof(ModPerfect) };
     }
 
     public abstract class ModPerfect : ModSuddenDeath
     {
-        public override Mods Name => Mods.Perfect;
-        public override FontAwesome Icon => FontAwesome.fa_close;
```
Comment (huoyaoyuan) [inline osu.Game/Modes/Mod.cs]: Why this icon forgot?

Comment (peppy) [inline osu.Game/Modes/Mod.cs]: unsure, need to update the font

**Comment by smoogipooo** [comment]
Apart from my comment, looks good.

File: osu.Game.Modes.Osu/OsuMod.cs [hunk 2]
```diff
@@ -56,41 +58,43 @@ public class OsuModNightcore : ModNightcore
...
+        public override string Name => "Autopilot";
         public override FontAwesome Icon => FontAwesome.fa_osu_mod_autopilot;
         public override string Description => @"Automatic cursor movement - just follow the rhythm.";
         public override double ScoreMultiplier => 0;
         public override bool Ranked => false;
-        public override Mods[] DisablesMods => new Mods[] { Mods.SpunOut, Mods.Relax, Mods.SuddenDeath, Mods.Perfect, Mods.NoFail, Mods.Autoplay, Mods.Cinema };
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModSpunOut), typeof(ModRelax), typeof(ModSuddenDeath), typeof(ModPerfect), typeof(ModNoFail), typeof(ModAutoplay), typeof(ModCinema) };
+    }
+
+    public class OsuModeAutoplay : ModAutoplay
```
Comment (smoogipooo) [inline osu.Game.Modes.Osu/OsuMod.cs]: Mode -> Mod

File: osu.Game/Modes/Mod.cs [hunk 3]
```diff
@@ -35,234 +35,129 @@ public abstract class Mod
...
     public abstract class ModEasy : Mod
     {
-        public override Mods Name => Mods.Easy;
+        public override string Name => "Easy";
         public override FontAwesome Icon => FontAwesome.fa_osu_mod_easy;
-        public override string Description => @"Reduces overall difficulty - larger circles, more forgiving HP drain, less accuracy required.";
+        public override string Description => "Reduces overall difficulty - larger circles, more forgiving HP drain, less accuracy required.";
         public override double ScoreMultiplier => 0.5;
         public override bool Ranked => true;
-        public override Mods[] DisablesMods => new Mods[] { Mods.HardRock };
```
Comment (huoyaoyuan) [inline osu.Game/Modes/Mod.cs]: Forgot `IncompatibleMods` here.

**Comment by huoyaoyuan** [comment]
And due to subclass checking, base mod and inherited mod don't need to appear in the same incompatible list.

File: osu.Game/Modes/Mod.cs [hunk 4]
```diff
@@ -35,234 +35,129 @@ public abstract class Mod
...
 
     public class ModCinema : ModAutoplay
     {
-        public override Mods Name => Mods.Cinema;
+        public override string Name => "Cinema";
         public override FontAwesome Icon => FontAwesome.fa_osu_mod_cinema;
     }
 
-    [Flags]
-    public enum Mods
```
Comment (Tom94) [inline osu.Game/Modes/Mod.cs]: We still need to be able to combine mods into a set of flags for serialization / database. Can you add a helper function that does that by iterating over a list of mods?

Comment (peppy) [inline osu.Game/Modes/Mod.cs]: when it's necessary, sure.

**Pull request merged**